packages/eslint-mdx/src/helpers.ts

```diff
@@ -161,8 +131,88 @@ export const arrayify = <T, R = T extends Array<infer S> ? S : T>(
     return arr
   }, [])
 
-// eslint-disable-next-line @typescript-eslint/prefer-optional-chain -- test coverage
-export const first = <T>(items: T[] | readonly T[]) => items && items[0]
+/**
+ * ! copied from https://github.com/just-jeb/angular-builders/blob/master/packages/custom-webpack/src/utils.ts#L53-L67
+ *
+ * This uses a dynamic import to load a module which may be ESM.
+ * CommonJS code can load ESM code via a dynamic import. Unfortunately, TypeScript
+ * will currently, unconditionally downlevel dynamic import into a require call.
+ * require calls cannot load ESM code and will result in a runtime error. To workaround
+ * this, a Function constructor is used to prevent TypeScript from changing the dynamic import.
+ * Once TypeScript provides support for keeping the dynamic import this workaround can
+ * be dropped.
+ *
+ * @param modulePath The path of the module to load.
+ * @returns A Promise that resolves to the dynamically imported module.
+ */
+/* istanbul ignore next */
+export const loadEsmModule = <T>(modulePath: URL | string): Promise<T> =>
```

Comment:
Can you replace all of this with https://github.com/wooorm/import-meta-resolve?

Reply:
`import-meta-resolve` will not solve that problem. It is also itself ESM, so that would be a problem here.
**But**, it’s much better at loading ESM/CJS because it supports export maps and other modern resolving features.

Reply:
So, is there anything you are proposing to change? I'm still confusing what `import-meta-resolve` will help with in this project. I believe native `import()` will handle _export maps and other modern resolving features_ already. We're loading the module instead of resolving its absolute path here.

Reply:
If it is possible to use `import-meta-resolve`, I recommend doing so.

`import()` resolves from `eslint-mdx/src/helpers.ts` and imports the resolved path. (and: in your code you first try to require them. That works okay but isn’t perfect.)

`import-meta-resolve` resolves modules *from wherever*

Reply:
`URL` is still available. In Node.js, you can create URLs from file paths, and file paths from URLs. `import-meta-resolve`/`import.meta.resolve` does not depend on `import.meta.url`.

But again: `import-meta-resolve` is ESM only. So if you are not in ESM here, you can’t easily use it. Though, you can create a small bundle of it:

```sh
npx esbuild import-meta-resolve --bundle --platform=node --outfile=vendor/import-meta-resolve.cjs`
```